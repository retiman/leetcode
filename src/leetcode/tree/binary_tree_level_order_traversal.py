# DIFFICULTY: MEDIUM
# ------------------
#
# Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right,
# level by level).
#
# See https://leetcode.com/problems/binary-tree-level-order-traversal
from collections import deque
from leetcode.tree.common.tree_node import TreeNode


class Solution:
    def levelOrder(self, root: TreeNode | None) -> list[list[int]]:
        """
        SOLUTION
        --------

        This is essentially a BFS algorithm from the root node.

        COMPLEXITY
        ----------

        Each node is visited once, so the time complexity is O(n).
        """
        if not root:
            return []

        result: list[list[int]] = []
        queue: deque[TreeNode] = deque([root])

        # Consume the nodes in the current level while noting the frontier nodes.  The nodes at the current level will
        # be recorded, then the frontier nodes are added to the queue.
        while queue:
            level: list[int] = []
            frontier: list[TreeNode] = []

            # Consume all the nodes from the current level and record them.  Don't use the queue.length in the for loop
            # as we are going to be modifying the queue during the loop.
            for _ in range(len(queue)):
                node = queue.popleft()
                level.append(node.val)

                if node.left:
                    frontier.append(node.left)
                if node.right:
                    frontier.append(node.right)

            # Add the recorded nodes at this level to the result.
            result.append(level)

            # Continue processing frontier nodes.
            queue.extend(frontier)

        return result
