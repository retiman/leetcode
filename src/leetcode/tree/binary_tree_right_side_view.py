# DIFFICULTY: MEDIUM
# ------------------
#
# Given the root of a binary tree, imagine yourself standing on the right side of it, return the values of the nodes
# you can see ordered from top to bottom.
#
# See https://leetcode.com/problems/binary-tree-right-side-view
from leetcode.tree.common.tree_node import TreeNode


class Solution:
    def rightSideView(self, root: TreeNode | None) -> list[int]:
        """
        SOLUTION
        --------

        We can use a simple DFS to traverse the tree, and prioritize the right child over the left child.  At each
        depth, we can add the first rightmost node to the result array, since that's the node we'll see when viewing it
        from the right.

        COMPLEXITY
        ----------

        Time complexity is O(n) because we are visiting each node once.

        Space complexity is O(n) because we are storing the nodes in a list.
        """
        depths: list[int] = []

        def dfs(node: TreeNode | None, depth: int) -> None:
            if not node:
                return

            # If the current depth is the lowest depth we've seen, then it's the rightmost node at this depth, because
            # we will always prioritize going down the right node.
            if depth == len(depths):
                depths.append(node.val)

            # Prioritize going down the right subtree first!
            dfs(node.right, depth + 1)
            dfs(node.left, depth + 1)

        dfs(root, 0)
        return depths
