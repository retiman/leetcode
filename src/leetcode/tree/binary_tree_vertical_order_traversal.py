# DIFFICULTY: MEDIUM
# ------------------
#
# Given the root of a binary tree, return the vertical order traversal of its nodes' values. (i.e., from top to bottom,
# column by column).
#
# If two nodes are in the same row and column, the order should be from left to right.
#
# See https://leetcode.com/problems/binary-tree-vertical-order-traversal
from collections import defaultdict, deque
from leetcode.tree.common.tree_node import TreeNode


class Solution:
    def verticalOrder(self, root: TreeNode | None) -> list[list[int]]:
        """
        SOLUTION
        --------

        There is no straightforward vertical order traversal.  However, a standard level order traversal can be used to
        get part of the way there.  If we assume that the root is at row 0 and column 0, we can do some bookkeeping
        while we traverse the nodes to assign a coordinate to every single node.

        While doing the traversal, we can keep track of a Map<Column, TreeNode[]> where each column is mapped to an
        array of nodes in the order they were encountered.  By doing a level order traversal (or BFS), we can ensure
        that nodes in the same "row" are encountered from left to right.

        COMPLEXITY
        ----------

        Time complexity is O(n) where n is the number of nodes.  The result computation at the end will run in O(k)
        time where k is the number of columns.  However, you're always guaranteed to have more nodes than columns (or
        the same) so overall time complexity is still O(n).

        Space complexity is O(n) because we are storing k columns mapping to lists of nodes, but the overall number of
        nodes stored in the map is at most n.  We are also using a queue, which has at most n nodes in it.
        """
        if not root:
            return []

        # Keep a map of column to nodes values.  Note that we don't actually need to store their row values, because if
        # we visit the nodes in the right order, the array will already be sorted by row.  The key is to push the left
        # child before the right child.
        mapping: dict[int, list[int]] = defaultdict(list)

        # We DO need to store the column values in our BFS queue though.  That's because as we shift items off the
        # queue, we will need to know which column it's in, so we can add it to the TreeNode array in the map above.
        queue: deque[tuple[TreeNode, int]] = deque([(root, 0)])

        # Finally, we have to keep track of the min and max columns so we know where to start with the vertical order
        # list.
        min_col = 0
        max_col = 0

        # The rest of the algorithm is a standard BFS.  We don't need to maintain a visited set because the nodes are
        # guaranteed to be organized as a tree, so we can't visit the same node twice.
        while queue:
            (node, col) = queue.popleft()

            # Add the node to this map for bookkeeping.
            mapping[col].append(node.val)

            # Update min/max columns.
            min_col = min(min_col, col)
            max_col = max(max_col, col)

            # Enqueue the children as usual in BFS, but make sure to enqueue the LEFT child first to keep our proper row
            # order.
            if node.left:
                queue.append((node.left, col - 1))
            if node.right:
                queue.append((node.right, col + 1))

        # Now we have a map of Column -> TreeNode[], and for each column, we just print out the values in the row.
        result: list[list[int]] = []
        for col in range(min_col, max_col + 1):
            # These node values are already in row order because we visited the left child before the right.
            values = mapping[col]
            result.append(values)

        return result
