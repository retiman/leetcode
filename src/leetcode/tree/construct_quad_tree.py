# DIFFICULTY: HARD
# ----------------
#
# Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
#
# Return the root of the Quad-Tree representing grid.
#
# See https://leetcode.com/problems/construct-quad-tree
from leetcode.tree.common.quad_tree import Node


class Solution:
    def construct(self, grid: list[list[int]]) -> Node | None:
        """
        SOLUTION
        --------

        To make a quad tree, recursively divide the matrix into quadrants until each quadrant is all 1's or all 0's.

        COMPLEXITY
        ----------

        Time complexity is O(n^2) because we are visiting each cell in the grid.

        Space complexity is O(log n) because the recursion depth is limited by the size of the grid.
        """
        if not grid or not grid[0]:
            return None

        return self.__construct(grid, 0, 0, len(grid))

    def __construct(self, grid: list[list[int]], row: int, column: int, size: int) -> Node | None:
        node = Node()

        if self.__isUniform(grid, row, column, size):
            node.val = bool(grid[row][column])
            node.isLeaf = True
            return node

        # If the grid segment here isn't uniform, we'll need to create 4 quadrants and construct nodes out of all 4
        # of them.
        #
        # Setting the `node.val` of this node is irrelevant as it's not a leaf.  We can just use the default value.
        half = size // 2
        node.isLeaf = False
        node.topLeft = self.__construct(grid, row, column, half)
        node.topRight = self.__construct(grid, row, column + half, half)
        node.bottomLeft = self.__construct(grid, row + half, column, half)
        node.bottomRight = self.__construct(grid, row + half, column + half, half)
        return node

    # Checks if a quadrant of the grid starting at [row, column] is uniformly all 1's or all 0's.
    def __isUniform(self, grid: list[list[int]], row: int, column: int, size: int) -> bool:
        value = grid[row][column]
        for i in range(row, row + size):
            for j in range(column, column + size):
                if grid[i][j] != value:
                    return False

        return True
