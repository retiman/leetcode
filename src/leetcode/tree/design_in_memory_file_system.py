# DIFFICULTY: HARD
# ----------------
#
# Design a data structure that simulates an in-memory file system.
#
# Implement the FileSystem class:
#
# - FileSystem()
#   Initializes the object of the system.
# - List<String> ls(String path)
#   If path is a file path, returns a list that only contains this file's name.
#   If path is a directory path, returns the list of file and directory names in this directory.
#   The answer should in lexicographic order.
# - void mkdir(String path)
#   Makes a new directory according to the given path. The given directory path does not exist.
#   If the middle directories in the path do not exist, you should create them as well.
# - void addContentToFile(String filePath, String content)
#   If filePath does not exist, creates that file containing given content.
#   If filePath already exists, appends the given content to original content.
# - String readContentFromFile(String filePath)
#   Returns the content in the file at filePath.
#
# See https://leetcode.com/problems/design-in-memory-file-system
class FileNode:
    def __init__(self):
        self.is_file = False
        self.content = ""
        self.children: "dict[str, FileNode]" = {}


class FileSystem:
    def __init__(self):
        """
        SOLUTION
        --------

        The problem states that we can assume all inputs are valid, so we don't need to check if we try to mkdir on path
        parts that contain files, or otherwise list children of files.  We can also assume that all listed directories
        and files exist.

        COMPLEXITY
        ----------

        Time complexity is O(n) for all operations.

        Space complexity is O(n) where n is the number of directories and files in the file system.
        """
        self.root = FileNode()

    def ls(self, path: str) -> list[str]:
        node = self.root

        normalized = self.__normalize(path)
        for part in normalized.split("/"):
            if part not in node.children:
                return []

            node = node.children[part]

            # Note that the ls function only wants to know the name of the file, not the absolute path, so don't bother
            # prefixing the parent path.
            if node.is_file:
                return [part]

        children = list(node.children.keys())
        children.sort()
        return children

    def mkdir(self, path: str) -> None:
        node = self.root

        normalized = self.__normalize(path)
        for part in normalized.split("/"):
            if part not in node.children:
                node.children[part] = FileNode()

            node = node.children[part]

    def addContentToFile(self, filePath: str, content: str) -> None:
        node = self.root

        normalized = self.__normalize(filePath)
        parts = normalized.split("/")
        for i, part in enumerate(parts):
            # The problem isn't exactly clear here, but the assumption is that we also mkdir the paths if they do not
            # exist.
            if part not in node.children:
                node.children[part] = FileNode()

            node = node.children[part]

            # The very last part is the filename, so append the content.
            if i == len(parts) - 1:
                node.content = node.content + content if node.content else content
                node.is_file = True

    def readContentFromFile(self, filePath: str) -> str:
        node = self.root

        normalized = self.__normalize(filePath)
        for part in normalized.split("/"):
            node = node.children[part]

        return node.content if node.content else ""

    def __normalize(self, path: str) -> str:
        return path[0:-1] if path[-1] == "/" else path
