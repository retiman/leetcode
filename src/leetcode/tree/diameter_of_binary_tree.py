# DIFFICULTY: EASY
# ----------------
#
# Given the root of a binary tree, return the length of the diameter of the tree.
#
# The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or
# may not pass through the root.
#
# The length of a path between two nodes is represented by the number of edges between them.
#
# See https://leetcode.com/problems/diameter-of-binary-tree
from leetcode.tree.common.tree_node import TreeNode


class Solution:
    def diameterOfBinaryTree(self, root: TreeNode | None) -> int:
        """
        SOLUTION
        --------

        For any node, the longest path might:

        - Pass through the node, in which case the longest path is the longest path of the left and right subtrees, plus
          one.
        - Not pass through the node, in which case the longest path is the max of the longest path of the left and
          right.

        We can use DFS to traverse the tree and keep track of the longest path we've seen so far.

        COMPLEXITY
        ----------

        Time complexity is O(n) where n is the number of nodes in the tree.

        Space complexity is O(n) because of the recursive stack.
        """
        max_path = 0

        def dfs(node: TreeNode | None) -> int:
            nonlocal max_path

            if not node:
                return 0

            # Figure out the max depth of the left and right subtrees.
            left = dfs(node.left)
            right = dfs(node.right)

            # The left and right depths represent the current diameter of the tree rooted at this node.  We'll compare
            # it with the global max diameter.
            current_path = left + right
            max_path = max(max_path, current_path)

            # Return the max depth starting from the current node.  Whichever path (left or right) is longer will be the
            # max depth, but we'll add one to account for the current node.
            return max(left, right) + 1

        dfs(root)
        return max_path
