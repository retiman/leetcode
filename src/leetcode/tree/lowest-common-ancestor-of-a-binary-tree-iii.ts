// DIFFICULTY: MEDIUM
//
// Given two nodes of a binary tree p and q, return their lowest common ancestor (LCA).
//
// According to the definition of LCA on Wikipedia: "The lowest common ancestor of two nodes p and q in a tree T is the
// lowest node that has both p and q as descendants (where we allow a node to be a descendant of itself)."
//
// See {@link https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree-iii}
import { _Node } from './common/parent-node';
export { lowestCommonAncestor };

// SOLUTION:
//
// Note that this problem is quite simple since we are given parent nodes.  All we have to do is traverse up to the root
// for one node and collect ancestors.  Then we traverse up to the root for the other node and check if any parent is in
// the ancestor set.  If it is, then we have found the lowest common ancestor.
//
// However, that solution involves the use of a set.  We can optimize this by using the two pointer technique in a
// clever way.  Let's consider paths from p and q to the LCA.
//
// If path[p] === path[q], then all we have to do is advance both pointers one step at a time and they will converge at
// the LCA eventually.
//
// If path[p] !== path[q], then one of the pointers will reach past the root and become null first.
//
// The goal, then is to make BOTH pointers reach the LCA at the same time.  So let's say path[p].length === 5, and that
// path[q].length === 10.  If we could craft an extended path for both p and q so that they both travel the same
// distance, and end up at the LCA, then that would work.
//
// To create this extended path, envision an extended path2[p] that goes from p to the root then to q then to the root
// (yes, imagine there's a direct path from root -> q).  Likewise, envision an extended path2[q] that goes from q to the
// root then from p to the root.
//
// Now if you make both the p and q pointers follow this path, they will both advance PAST the root, but then eventually
// converge at the LCA before they hit the root again.
//
// COMPLEXITY:
//
// For the solution involving a set, the time complexity is O(h + h) where h is the depth of the deeper node.  The
// space complexity is O(h) to store the ancestors.
//
// For the solution not involving a set, the time complexity is O(h + h) where h is the depth of the deeper node.
function lowestCommonAncestor(p: _Node | null, q: _Node | null): _Node | null {
  // Create new pointers to advance; we need the original starting points p and q.
  let [a, b] = [p, q];

  // Eventually a and b will converge at the LCA.
  while (a !== b) {
    // Advance a = p -> root -> q -> root.  Pointers converge before hitting the root again.
    a = a !== null ? a.parent : q;

    // Advance b = q -> root -> p -> root.  Pointers converge before hitting the root again.
    b = b !== null ? b.parent : p;
  }

  // Return either pointer; they are both at the LCA.
  return a;
}
