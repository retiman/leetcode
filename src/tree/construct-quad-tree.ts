// DIFFICULTY: Hard
//
// Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
//
// Return the root of the Quad-Tree representing grid.
//
// See https://leetcode.com/problems/construct-quad-tree/
// See https://en.wikipedia.org/wiki/Quadtree
import { _Node } from './common/quad-tree';
export { construct };

// SOLUTION:
//
// To make a quad tree, recursively divide the matrix into quadrants until each quadrant is all 1's or all 0's.
function construct(_grid: number[][]): _Node | null {
  if (_grid.length === 0 || _grid[0].length === 0) {
    return null;
  }

  return constructInternal(_grid, 0, 0, _grid.length);
}

function constructInternal(grid: number[][], row: number, column: number, size: number): _Node {
  const node = new _Node();

  // If the grid segment is uniform, we can create a single quad tree node and set the value to all 1's or 0's.
  if (isUniform(grid, row, column, size)) {
    node.val = grid[row][column] === 1;
    node.isLeaf = true;
    return node;
  }

  // If the grid segment here isn't uniform, we'll need to create 4 quadrants and construct nodes out of all 4 of
  // them.
  //
  // Setting the `node.val` of this node is irrelevant as it's not a leaf.  We can just use the default value.
  const half = size / 2;
  node.isLeaf = false;
  node.topLeft = constructInternal(grid, row, column, half);
  node.topRight = constructInternal(grid, row, column + half, half);
  node.bottomLeft = constructInternal(grid, row + half, column, half);
  node.bottomRight = constructInternal(grid, row + half, column + half, half);

  return node;
}

// Checks if a quadrant of the grid starting at [row, column] is uniformly all 1's or all 0's.
function isUniform(grid: number[][], row: number, column: number, size: number) {
  const value = grid[row][column];
  for (let i = row; i < row + size; i++) {
    for (let j = column; j < column + size; j++) {
      if (grid[i][j] !== value) {
        return false;
      }
    }
  }

  return true;
}
