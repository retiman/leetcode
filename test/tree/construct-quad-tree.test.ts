// DIFFICULTY: Hard
//
// Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
//
// Return the root of the Quad-Tree representing grid.
//
// See https://leetcode.com/problems/construct-quad-tree/
// See https://en.wikipedia.org/wiki/Quadtree
import { _Node } from '../../src/tree/construct-quad-tree';

describe('construct quad tree', () => {
  function construct(_grid: number[][]): _Node | null {
    return null;
  }

  function __serialize(node: _Node) {
    
  }

  test('test case 1', async () => {
    const grid = [
      [0, 1],
      [1, 0]
    ];
    expect(construct(grid)).toBeNull();
  });
});
