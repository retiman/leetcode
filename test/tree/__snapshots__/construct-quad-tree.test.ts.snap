// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`construct quad tree test case 1 1`] = `
_Node {
  "bottomLeft": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": true,
  },
  "bottomRight": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": false,
  },
  "isLeaf": false,
  "topLeft": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": false,
  },
  "topRight": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": true,
  },
  "val": false,
}
`;

exports[`construct quad tree test case 2 1`] = `
_Node {
  "bottomLeft": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": true,
  },
  "bottomRight": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": false,
  },
  "isLeaf": false,
  "topLeft": _Node {
    "bottomLeft": null,
    "bottomRight": null,
    "isLeaf": true,
    "topLeft": null,
    "topRight": null,
    "val": true,
  },
  "topRight": _Node {
    "bottomLeft": _Node {
      "bottomLeft": null,
      "bottomRight": null,
      "isLeaf": true,
      "topLeft": null,
      "topRight": null,
      "val": true,
    },
    "bottomRight": _Node {
      "bottomLeft": null,
      "bottomRight": null,
      "isLeaf": true,
      "topLeft": null,
      "topRight": null,
      "val": true,
    },
    "isLeaf": false,
    "topLeft": _Node {
      "bottomLeft": null,
      "bottomRight": null,
      "isLeaf": true,
      "topLeft": null,
      "topRight": null,
      "val": false,
    },
    "topRight": _Node {
      "bottomLeft": null,
      "bottomRight": null,
      "isLeaf": true,
      "topLeft": null,
      "topRight": null,
      "val": false,
    },
    "val": false,
  },
  "val": false,
}
`;
